pkg/osbuild/fips.go

```diff
@@ -0,0 +1,53 @@
+package osbuild
+
+import (
+	"os"
+
+	"github.com/osbuild/images/internal/common"
+	"github.com/osbuild/images/internal/fsnode"
+	"github.com/osbuild/images/pkg/disk"
+)
+
+func GenFIPSKernelOptions(pt *disk.PartitionTable) []string {
+	cmdline := make([]string, 0)
+	cmdline = append(cmdline, "fips=1")
+	if bootMnt := pt.FindMountable("/boot"); bootMnt != nil {
+		boot := bootMnt.GetFSSpec()
+		if label := boot.Label; label != "" {
+			karg := "boot=LABEL=" + label
+			cmdline = append(cmdline, karg)
+		} else if uuid := boot.UUID; uuid != "" {
+			karg := "boot=UUID=" + uuid
+			cmdline = append(cmdline, karg)
+		}
+	}
+	return cmdline
+}
+
+func GenFIPSStages() (stages []*Stage) {
+	system_fips_file, _ := fsnode.NewFile(
+		"/etc/system-fips",
+		common.ToPtr(os.FileMode(0644)),
+		"root",
+		"root",
+		[]byte("# FIPS module installation complete\n"))
+
+	files := []*fsnode.File{system_fips_file}
+
+	stages = []*Stage{
+		NewUpdateCryptoPoliciesStage(
+			&UpdateCryptoPoliciesStageOptions{
+				Policy: "FIPS",
+			}),
+		NewDracutConfStage(
+			&DracutConfStageOptions{
+				Filename: "40-fips.conf",
+				Config: DracutConfigFile{
+					AddModules: []string{"fips"},
+				},
+			}),
+	}
+	stages = append(stages, GenFileNodesStages(files)...)
```

Comment:
This won't work, since the `GenFileNodesStages()` generates only stages, but no inline sources, for the file. These need to be added to the final manifest, so that the generate stages can copy them to the image FS tree.

I suggest to modify the function to return `([]*Stage, []*fsnode.File)` and then add those stages to the pipeline and the files to the custom files in the pipeline implementation:
https://github.com/osbuild/images/blob/63a1eead26a7c802dbcebe863439f591be6dc6e5/pkg/manifest/ostree_deployment.go#L52
https://github.com/osbuild/images/blob/63a1eead26a7c802dbcebe863439f591be6dc6e5/pkg/manifest/os.go#L133

The existing pipeline implementation will then handle the generation of necessary stages to copy the file into the image and set its attributes (this also ensures that instead of multiple instances of those stages, there will be only one handling all custom files). The following code then handles the creation of inline sources for custom files:
https://github.com/osbuild/images/blob/63a1eead26a7c802dbcebe863439f591be6dc6e5/pkg/manifest/os.go#L817-L820
https://github.com/osbuild/images/blob/63a1eead26a7c802dbcebe863439f591be6dc6e5/pkg/manifest/ostree_deployment.go#L362-L365

Reply:
Yeah, you are right, but doesn't this complicate things a lot in comparison with the previous implementation? I personally prefer the previous approach.

Reply:
Actually, If I do that the file stages will be appended twice to the pipepline unless I remove the file stages from the `GenFIPSStages()` function. That doesn't make much sense to me honestly. Having the stages and files generation in two different functions feels more natural to me.

Reply:
Reverted to previous implementation

Reply:
> Actually, If I do that the file stages will be appended twice to the pipepline unless I remove the file stages from the `GenFIPSStages()` function. That doesn't make much sense to me honestly. Having the stages and files generation in two different functions feels more natural to me.

You are right, I didn't realize that...